### PR TITLE
Add playAll() and playSync()

### PR DESCRIPTION
**Pull request by moos**
Attempt at #228.   See example for usage.

My use case is: "X minutes, Y seconds" which is comprised of 4 files (X.mp3, Y.mp3, minutes.mp3, seconds.mp3) and X & Y are numeric between 0 to 60, requiring only 62 files vs. 3600! if spoken as one sentence.

Possible additions:
[ ] Add a `loop: true/false` parameter
[ ] others?...

### CODE REVIEW

File: lib/audio_cache.dart [hunk 1]
```diff
@@ -81,8 +81,29 @@ class AudioCache {
     return loadedFiles[fileName];
   }
 
-  AudioPlayer _player(PlayerMode mode) {
-    return fixedPlayer ?? new AudioPlayer(mode: mode);
+  AudioPlayer _player(PlayerMode mode, [localInstance]) {
+    return localInstance ?? fixedPlayer ?? new AudioPlayer(mode: mode);
+  }
+
+  Future<AudioPlayer> _play({
```
Comment (moos) [inline lib/audio_cache.dart]: Internal helper function to allow re-use of player instance.

File: lib/audio_cache.dart [hunk 2]
```diff
@@ -96,15 +117,92 @@ class AudioCache {
...
+      c.complete(playerInstance);
+    });
+
+    return c.future;
+  }
+
+  /// Like [play], but plays multiple files in sequence.
+  ///
+  /// The instance is returned immediately, to allow later access like pausing and resuming.
+  Future<AudioPlayer> playAll(List<String> fileNames,
```
Comment (moos) [inline lib/audio_cache.dart]: Certainly open to naming suggestion -- it is a bit different than `loadAll()` which returns a Futures List -- this returns a single player instance future.

File: lib/audio_cache.dart [hunk 3]
```diff
@@ -96,15 +117,92 @@ class AudioCache {
...
+        // force mode to get completion event
+        mode: PlayerMode.MEDIA_PLAYER,
+        stayAwake: stayAwake,
+        // re-use same player instance
+        player: playerInstance,
+      );
+    };
+
+    // pre-load all files
+    await loadAll(fileNames);
```
Comment (moos) [inline lib/audio_cache.dart]: Possible optimization to pre-load and start playing the first file, then load the rest.  May be it's overly complicating it -- sequential is really meant for shorter files.

File: example/lib/main.dart [hunk 4]
```diff
@@ -99,12 +99,20 @@ class _ExampleAppState extends State<ExampleApp> {
       _btn('Play', () => audioCache.play('audio.mp3')),
       Text('Loop Local Asset \'audio.mp3\':'),
       _btn('Loop', () => audioCache.loop('audio.mp3')),
-      Text('Play Local Asset \'audio2.mp3\':'),
-      _btn('Play', () => audioCache.play('audio2.mp3')),
```
Comment (juanmanual) [inline example/lib/main.dart]: Why remove these files? Just to make room for the new buttons?

Comment (moos) [reply example/lib/main.dart]: Yes -- also they seemed redundant, just playing a different file -- unless I missed something.

**Comment by moos** [comment]
Let me know if this is going in the right direction -- I can work on updating the readme and other docs.

Also, what do you think of renaming `playAll()` to `playList()`?  In fact it'd be great if the list was made available so it could be added to (or removed from).
```
p =  player.playList(['foo', 'bar']);
// little later
p.playlist.add('baz');
```
Also nice-to-have is to know when the list has finished.  So maybe an `onPlayListCompletion` !?

**Comment by danielR2001** [comment]
hi, can you also add to your feature loading next songs (with choose option of how much songs to load ahead).

**Comment by danielR2001** [comment]
You will need multiple audio players and just use set URL on each and when you want to play next use resume.

File: lib/audio_cache.dart [hunk 5]
```diff
@@ -85,6 +86,28 @@ class AudioCache {
     return fixedPlayer ?? new AudioPlayer(mode: mode);
   }
 
+  @visibleForTesting
+  Future<AudioPlayer> inner_play({
```
Comment (luanpotter) [inline lib/audio_cache.dart]: camelCase for Dart code

File: lib/audio_cache.dart [hunk 6]
```diff
@@ -85,6 +86,28 @@ class AudioCache {
...
+      File file,        // either file object
+      String fileName,  // or file name
+      double volume = 1.0,
+      bool isNotification,
+      PlayerMode mode = PlayerMode.MEDIA_PLAYER,
+      bool stayAwake,
+      AudioPlayer player}) async {
```
Comment (luanpotter) [inline lib/audio_cache.dart]: add trailing comma here

File: lib/audio_cache.dart [hunk 7]
```diff
@@ -85,6 +86,28 @@ class AudioCache {
...
+
+    if (file == null) file = await load(fileName);
+    if (player == null) player = _player(mode);
+    await player.play(
```
Comment (luanpotter) [inline lib/audio_cache.dart]: remove extra empty line here

Comment (luanpotter) [reply lib/audio_cache.dart]: use the ?? operator

File: lib/audio_cache.dart [hunk 8]
```diff
@@ -96,15 +119,77 @@ class AudioCache {
...
+    return c.future;
+  }
+
+  /// Like [play], but plays multiple files in sequence.
+  ///
+  /// The instance is returned immediately, to allow later access like pausing and resuming.
+  Future<AudioPlayer> playAll(List<String> fileNames,
+      {double volume = 1.0,
+      bool isNotification,
+      bool stayAwake}) async {
```
Comment (luanpotter) [inline lib/audio_cache.dart]: trailing comma

**Comment by fdietze** [comment]
I like `playSequentially` or similar more than `playAll`.

File: lib/audio_cache.dart [hunk 9]
```diff
@@ -96,15 +119,77 @@ class AudioCache {
...
+    if (fileNames.length == 0) return new Future.value(null);
+    AudioPlayer playerInstance;
+    StreamSubscription completionListener;
+
+    // inner function
+    playNext() async {
+      playerInstance = await inner_play(
+        fileName: fileNames.removeAt(0),
```
Comment (fdietze) [inline lib/audio_cache.dart:169]: If possible, please avoid mutating the argument. We do not know what the caller wants to do with the list afterwards.

**Comment by erickzanardo** [comment]
This PR is quite outdated and haven't got updates in a long time. Closing this. If the author wants to continue working on this, just lets us know and we can re open the PR.

**Pull request closed without merging**